### DISCUSSION TITLE
쿠버네티스 사용해야하나?

### DISCUSSION THREAD

**Discussion by sokojh** [heading1/WYLSBingsu#60] [General]
# 서론
이번 프로젝트를 진행하면서, 쿠버네티스(k8s)를 사용 해보고 싶은 생각이 들었습니다.
그래서 6월 23일 하루를 온전히 사용하며 쿠버네티스와 도커의 개념과 실전 use case를 찾아봤지만,

## 결론적으로 도커는 도입해도 좋지만,  쿠버네티스는...

먼저 쿠버네티스란 무엇인지 설명드리겠습니다.

쿠버네티스는 배의 조타라는 뜻의 그리스어로 즉 항해하는 키를 뜻하는 단어입니다.

이름에서도 알다시피 큰 배를 조종 한다는 뜻을 지녔는데, 바로 도커를 조종한다고 볼 수 있습니다.
<img width="500"
     src="https://user-images.githubusercontent.com/83265848/175263461-c05248d6-b2b3-43f9-945b-b5edb3913a12.png"
     alt="도커 혼란">
> 많아진 도커 컨테이너

도커는 앱을 이미지화 시켜서 빠르게 컨테이너로 띄우고 배포하는 것에 특화가 되어있지만,

우리가 일일이 확인해주고 컨테이너가 죽으면 다시 켜주고 손이 많이 가는 아이(?)라고 할 수 있습니다.

그래서 귀찮은것을 매우 싫어하는 개발자들은 고민 했습니다.

이걸 자동화 할 순 없을까?  MSA로 서비스를 하게 되면 컨테이너가 수백개까지 될 수 있는데, 

도스 앞에 앉아서 컨테이너가 꺼져있는지 살아있는지 확인하는 것은 매우 시간낭비라 생각했기 때문에,
<img width="500"
     src="https://user-images.githubusercontent.com/83265848/175264424-b99f6fc9-7ace-404d-90b7-d174cf03c4ec.png"
     alt="쿠버네티스 혼란 짤">

이런 점을 해결한 쿠버네티스가 나오게됩니다.

쿠버네티스는 도커 컨테이너를 자동으로 배포 시켜주며(서버 다운없이), 컨테이너가 죽으면 다시 재시작까지 해주게 됩니다.

즉, 컨테이너가 많더라도 우리의 손 없이도 자동적으로 감시 할 수 있게 해주는 도구가 쿠버네티스라고 할 수 있습니다.

# 그럼 여기서 의문 

그렇다면 우리 서비스에 쿠버네티스를 도입하는게 맞을까요?

그렇습니다. 다들 예상 하셨겠지만, 필요하지 않습니다.

저희는 마이크로 서비스를 운영하지도 않고 쿠버네티스를 도입할 만큼 많은 컨테이너를 동작시키지도 않습니다.

제 예상으로는 인스턴스 한개로도 운영 가능할 것 같다는 생각을 하고 있습니다.

기술은 목적이 확실해야 하다고 생각합니다. 닭잡는데 소잡는 칼이 필요없듯이 적재적소에 맞는 기술을 써야 된다는 생각이듭니다.

팀원분들도 이 부분에 대해서 한번쯤 고심해보셨으면 좋겠습니다. 😃

**Comment by jinux127** [comment]
프론트이지만 항상 많이 듣는 키워드인 도커와 쿠버네티스 인데 지호님의 찰떡같은 설명으로 다시한번 이해가 가네용 ㅋㅋㅋ 기회가 된다면 해보면 좋겠지만 저희 프로젝트 같이 규모가 작은 경우엔 굳이 라는 생각이 들긴하네요.. 만약 적용하게 된다면 프론트쪽에는 무슨 작업을 해줘야 하나 궁금증도 생기네요 😄

**Reply by sokojh** [reply]
MSA로 서비스를 전환하게된다면 사용해보는 것도 좋지만,  저희는 MSA로 진행하지는 않고 있으니 도커 컨테이너만으로도 충분히 커버가 가능할 것이라고 생각하고 있습니다. ㅎㅎ  
저도 정확한것은 아니지만, 만약 도커를 통해서 MSA시스템(쿠버네티스를 사용한)으로 바꾸게 된다면, 유저 API, 글 API 도메인이나 포트가 각각 다르므로 클라이언트 쪽에서는 라우터 설정을 해줘야 될 것 같습니다. 😄
